capsule_network.py

```diff
@@ -87,29 +89,14 @@ def margin_loss(self, input, target, size_average=True):
 
         return L_c
 
-    def reconstruction_loss(self, images, input, size_average=True):
-        # Get the lengths of capsule outputs.
-        v_mag = torch.sqrt((input**2).sum(dim=2))
-
-        # Get index of longest capsule output.
-        _, v_max_index = v_mag.max(dim=1)
-        v_max_index = v_max_index.data
-
-        # Use just the winning capsule's representation (and zeros for other capsules) to reconstruct input image.
-        batch_size = input.size(0)
-        all_masked = [None] * batch_size
-        for batch_idx in range(batch_size):
-            # Get one sample from the batch.
-            input_batch = input[batch_idx]
-
-            # Copy only the maximum capsule index from this batch sample.
-            # This masks out (leaves as zero) the other capsules in this sample.
-            batch_masked = Variable(torch.zeros(input_batch.size())).cuda()
-            batch_masked[v_max_index[batch_idx]] = input_batch[v_max_index[batch_idx]]
-            all_masked[batch_idx] = batch_masked
-
-        # Stack masked capsules over the batch dimension.
-        masked = torch.stack(all_masked, dim=0)
+    def reconstruction_loss(self, images, input, target, size_average=True):
+        # Use the target to reconstruct input image.
+        # (batch_size, num_output_units, output_unit_size)
+        input = torch.squeeze(input, 3)
+        # (batch_size, num_output_units, 1)
+        target = torch.unsqueeze(target, 2)
+        # (batch_size, output_unit_size, 1)
+        masked = torch.matmul(input.transpose(2,1), target)
```

Comment:
hmm, i don't quite understand this part.  what does 'masked' look like for a typical sample after this matmul?  also see earlier comment, i thought the input was all capsules with all but one masked out -- otherwise the reconstruction would be confused by different classes of digits

Reply:
First, there is no doubt to use the target to reconstruction. But my implementation maybe wrong after I rethink about this question. I thought the reconstruction network just need 16 values to reconstruct 10 images, but we know every digit capsule represent 1 entity(digit), so we should use all the 160 values to finish the reconstruction,  and this also prove the fact how to reconstruct 2 overlapped MNIST at a time as the original paper said(Although this can be done by find the two max longest digit capsule and then reconstruct them two times and sum it, not enough neat as aforementioned). So yes, I think you're right. :)
just change the masked part to ' masked = input * target ' and change the number of first layer units to ' num_output_units*output_unit_size ' then it will work.